### PR TITLE
Bugfixes to image morph and nearest neighbors

### LINKED ISSUE #1594: Change initial values of min and max to be +/- infinity
I think this was missed in the 3.4 release. 
As per  https://github.com/arrayfire/arrayfire/pull/1547#issuecomment-240919098

### LINKED ISSUE #1596: Nearest Neighbour Distance Calculation wrong results on CUDA and OpenCL Backend
There appears to be an issue in the nearest neighbour function in the CUDA and OpenCL Backend (CPU returns the correct results).
Here is a small Test Case Illustrating the issue:

```C++
TEST(NearestNeighbour, CUDA)
{
	af::setBackend(AF_BACKEND_CUDA);
	double train[1 * 2] = {
		5, 5,
	};

	double query[5 * 2] = {
		0, 0,
		3.5, 4,
		5, 5,
		6, 5,
		8, 6.5
	};
	array t(2, 1, train);
	array q(2, 5, query);
	array indices;
	array distances;
	nearestNeighbour(indices, distances, q, t, 0, 1, AF_SSD);

	double expectedDistances[5] = {
		(5 - 0) * (5 - 0) + (5 - 0) * (5 - 0),
		(5 - 3.5) * (5 - 3.5) + (5 - 4) * (5 - 4),
		(5 - 5) * (5 - 5) + (5 - 5) * (5 - 5),
		(5 - 6) * (5 - 6) + (5 - 5) * (5 - 5),
		(5 - 8) * (5 - 8) + (5 - 6.5) * (5 - 6.5)
	};

	double* actualDistances = distances.host<double>();
	for (int i = 0; i < 5; i++)
	{
		EXPECT_EQ(expectedDistances[i], actualDistances[i]);
	}
}
```
The formula for calculating the sum of square distance between Point p1 and p2 is: `(p1.x - p2.x)^2 + (p1.y - p2.y)^2`
When trying to calculate the distance of every point in query to (5 | 5) the CPU Backend returns the correct results (those in expectedDistances). Those are:
`   50.0000     3.2500     0.0000     1.0000    11.2500 `

However the CUDA and OpenCL backend return the following values:
`   50.0000    27.2500    25.0000    26.0000    34.0000 `

This issue does not only affect the `AF_SSD` match type, but also the `AF_SAD` (sum of absolute difference). Here the correct results would be: 
`10.0000     2.5000     0.0000     1.0000     4.5000 `
but instead are:
`   10.0000     6.5000     5.0000     6.0000     8.0000 `

I am using ArrayFire v3.4.0 installed with the binaries from your website.
```
ArrayFire v3.4.0 (CUDA, 64-bit Windows, build 10d9716)
Platform: CUDA Toolkit 7.5, Driver: CUDA Driver Version: 7050
[0] GeForce GT 730M, 1024 MB, CUDA Compute 3.5
```

### PR DESCRIPTION
**Pull request by pavanky**
- Fixes issues with erode and dilate at corner cases
- Fixes #1594 
- Fixes #1596

### CODE REVIEW

**Comment by pavanky** [comment]
@shehzan10 I am not making my PRs run tests run automatically. Can you launch and run the tests when possible

**Comment by shehzan10** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Comment by pavanky** [comment]
@shehzan10 you can kill old windows build. I missed something earlier by not testing OpenCL backend.

**Comment by pavanky** [comment]
@umar456 @shehzan10 
Ideally the `options` should be a daughter class of `stringstream` with customizations. This will get rid of hacks like the last commit I have here and other work arounds like (`toNum`).

May be the payoff is not that great, but that is something to consider going forward.

File: src/backend/cuda/math.hpp [hunk 1]
```diff
@@ -93,26 +93,30 @@ namespace cuda
...
-    template<> __device__  char   limit_max<char>()   { return 0x7f; }
-    template<> __device__  char   limit_min<char>()   { return 0x80; }
-    template<> __device__  float  limit_max<float>()  { return  CUDART_INF_F; }
-    template<> __device__  float  limit_min<float>()  { return -CUDART_INF_F; }
-    template<> __device__  double limit_max<double>() { return  CUDART_INF; }
-    template<> __device__  double limit_min<double>() { return -CUDART_INF; }
-    template<> __device__  short  limit_max<short>()  { return 0x7fff; }
-    template<> __device__  short  limit_min<short>()  { return 0x8000; }
-    template<> __device__  ushort limit_max<ushort>() { return ((ushort)1) << (8 * sizeof(ushort) - 1); }
+    template <typename T> __device__ T maxval() { return 1u << (8 * sizeof(T) - 1); }
```
Comment (umar456) [inline src/backend/cuda/math.hpp]: Hmm. I liked the old name

File: src/backend/cuda/kernel/nearest_neighbour.hpp [hunk 2]
```diff
@@ -392,26 +399,26 @@ __global__ void select_matches(
...
-                __syncthreads();
+    // Reduce best matches and find the best of them all
+    for (unsigned i = blockDim.y / 2; i > 0; i >>= 1) {
+        if (threadIdx.y < i) {
+            To dist = s_dist[sid + i];
+            if (dist < s_dist[sid]) {
+                s_dist[sid] = dist;
+                s_idx[sid]  = s_idx[sid + i];
             }
+            __syncthreads();
```
Comment (umar456) [inline src/backend/cuda/kernel/nearest_neighbour.hpp]: Isn't this still undefined?

**Comment by shehzan10** [comment]
build arrayfire windows ci

File: src/backend/cuda/kernel/nearest_neighbour.hpp (hunk 2 above)
Comment (pavanky) [inline src/backend/cuda/kernel/nearest_neighbour.hpp]: It was already like that (look at the left). I fixed it in one place but missed this. I'll fix it.

File: src/backend/cuda/math.hpp (hunk 1 above)
Comment (pavanky) [inline src/backend/cuda/math.hpp]: I did not want it to be confused with numeric limits. I could change it back if you feel strongly about.

Comment (umar456) [inline src/backend/cuda/math.hpp]: Isn't it doing the same thing as numeric limits. If anything it should have the same interface.

Comment (pavanky) [inline src/backend/cuda/math.hpp]: It has different behavior than numeric limits for floats and doubles. It had same behavior before, but this commit changed the behavior so I changed the name as well.

**Comment by shehzan10** [comment]
build arrayfire windows opencl ci

**Comment by pavanky** [comment]
@shehzan10 I missed the change in __syncthreads like @umar456  mentioned. Can you fix it in a different PR.

BTW, it would also be better if you guys start commenting this as reviews.
